controllers/perconaserverformysql_controller.go

```diff
@@ -191,23 +192,74 @@ func (r *PerconaServerForMySQLReconciler) reconcileDatabase(
 		return errors.Wrap(err, "get init image")
 	}
 
-	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.StatefulSet(cr, initImage), r.Scheme)
-	if err != nil {
+	if err := k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.StatefulSet(cr, initImage), r.Scheme); err != nil {
 		return errors.Wrap(err, "reconcile sts")
 	}
-	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.Service(cr), r.Scheme)
-	if err != nil {
-		return errors.Wrap(err, "reconcile svc")
+
+	if err := r.reconcileMySQLServices(ctx, cr); err != nil {
+		return errors.Wrap(err, "reconcile services")
 	}
-	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.PrimaryService(cr), r.Scheme)
-	if err != nil {
+
+	return nil
+}
+
+func (r *PerconaServerForMySQLReconciler) reconcileMySQLServices(
+	ctx context.Context,
+	cr *apiv2.PerconaServerForMySQL,
+) error {
+	if err := k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.PrimaryService(cr), r.Scheme); err != nil {
 		return errors.Wrap(err, "reconcile primary svc")
 	}
-	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.UnreadyService(cr), r.Scheme)
-	if err != nil {
+
+	if err := k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.UnreadyService(cr), r.Scheme); err != nil {
 		return errors.Wrap(err, "reconcile unready svc")
 	}
 
+	if err := k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.HeadlessService(cr), r.Scheme); err != nil {
+		return errors.Wrap(err, "reconcile headless svc")
+	}
+
+	pods, err := k8s.PodsByLabels(ctx, r.Client, mysql.MatchLabels(cr))
+	if err != nil {
+		return errors.Wrap(err, "get MySQL pods")
+	}
+
+	for _, pod := range pods {
```

Comment:
PSMDB practice should that LoadBalancers start is very slow, up to 5 minutes each.
it is better to have parallelization here - create services (according to `cr.Spec.MySQL.Size`) even before Pods.

Reply:
yes, LoadBalancer readiness is really slow because it waits for the external IP but creating the service doesn't wait for external IP. It creates and continues. In PSMDB we're deliberately waiting for the external IP to be ready.

Reply:
startup time is important metric 😞